tests/unit/other.tcl

```diff
@@ -634,39 +634,73 @@ start_server {tags {"other external:skip"} overrides {cluster-compatibility-samp
+    } {} {cluster:skip}
 
-        # Normal cross slot command
-        r mset foo bar bar foo ;# 1
-        r del foo bar ;# 2
+    test {Transaction is incompatible with cluster mode} {
+        set incompatible_ops [s cluster_incompatible_ops]
 
-        # Lua script
-        r eval {#!lua
-            redis.call('mset', KEYS[1], 0, KEYS[2], 0)
-        } 2 foo bar ;# 3
+        # Incomplete transaction
+        catch {r EXEC}
+        r multi
+        r exec
+        assert_equal $incompatible_ops [s cluster_incompatible_ops]
 
-        # Transaction, SET and DEL have keys with different slots, 4
+        # Transaction, SET and DEL have keys with different slots
         r multi
         r set foo bar
         r del bar
         r exec
+        assert_equal [expr $incompatible_ops + 1] [s cluster_incompatible_ops]
+    } {} {cluster:skip}
 
-        # Function call
-        r function flush
-        r function load {#!lua name=test
-            redis.register_function('ftest', function(KEYS, ARGV)
-                redis.call('mset', KEYS[1], 0, KEYS[2], 0)
-            end)
-        }
-        r fcall ftest 2 foo bar ;# 5
+    test {Lua scripts are incompatible with cluster mode} {
+        # Lua script, declared keys have different slots, it is not a compatible operation
+        set incompatible_ops [s cluster_incompatible_ops]
+        r eval {#!lua
+            redis.call('mset', KEYS[1], 0, KEYS[2], 0)
+        } 2 foo bar
+        assert_equal [expr $incompatible_ops + 1] [s cluster_incompatible_ops]
+
+        # Lua script, no declared keys, but accessing keys have different slots,
+        # it is not a compatible operation
+        set incompatible_ops [s cluster_incompatible_ops]
+        r eval {#!lua
+            redis.call('mset', 'foo', 0, 'bar', 0)
+        } 0
+        assert_equal [expr $incompatible_ops + 1] [s cluster_incompatible_ops]
+
+        # Lua script, declared keys have the same slot, but accessing keys
+        # have different slots in one command, even with flag 'allow-cross-slot-keys',
+        # it still is not a compatible operation
+        set incompatible_ops [s cluster_incompatible_ops]
+        r eval {#!lua flags=allow-cross-slot-keys
+            redis.call('mset', 'foo', 0, 'bar', 0)
+            redis.call('mset', KEYS[1], 0, KEYS[2], 0)
+        } 2 foo bar{foo}
+        assert_equal [expr $incompatible_ops + 1] [s cluster_incompatible_ops]
+
+        # Lua script, declared keys have the same slot, but accessing keys have different slots
+        # in multiple commands, and with flag 'allow-cross-slot-keys', it is a compatible operation
+        set incompatible_ops [s cluster_incompatible_ops]
+        r eval {#!lua flags=allow-cross-slot-keys
+            redis.call('set', 'foo', 0)
+            redis.call('set', 'bar', 0)
+            redis.call('mset', KEYS[1], 0, KEYS[2], 0)
+        } 2 foo bar{foo}
+        assert_equal $incompatible_ops [s cluster_incompatible_ops]
```

Comment:
shouldn't this have incremented the counter?

Reply:
for cluster mode, we don't report an error since we add `allow-cross-slot-keys` here, so i don't increment counter. 
but there may be a problem that we can not know if the node has the slot in standalone mode, i added comments https://github.com/redis/redis/pull/13846/files#diff-864bd15102480e5fabce95fc39c9b2a177afcbcb165eae2a7f71d890a7d28fe5R542-R544
I am not sure if we should consider it incompatible with cluster mode.

Reply:
ohh i missed it.
i thought it's a test that checks multiple single-key commands in a script with different slots.
i guess that's already covered.